### PR TITLE
Weighted vote

### PR DESCRIPTION
**Pull request by cryptoAtwill**
## Changes

Use weight, aka TokenAmount to check the voting.

## Tests

```
cargo test --all
```

### CODE REVIEW

File: gateway/src/cron.rs [hunk 1]
```diff
@@ -39,9 +44,13 @@ impl Validators {
         }
     }
 
-    /// Checks if an address is a validator
-    pub fn is_validator(&self, addr: &Address) -> bool {
-        self.validators.validators().iter().any(|x| x.addr == *addr)
+    /// Get the weight of a validator
+    pub fn get_weight(&self, addr: &Address) -> Option<TokenAmount> {
```
Comment (adlrocha) [inline gateway/src/cron.rs:48]:
```suggestion
    pub fn get_validator_weight(&self, addr: &Address) -> Option<TokenAmount> {
```

File: gateway/src/cron.rs [hunk 2]
```diff
@@ -131,11 +140,13 @@ impl CronSubmission {
         &mut self,
         store: &BS,
         submitter: Address,
+        submitter_weight: TokenAmount,
         checkpoint: CronCheckpoint,
-    ) -> anyhow::Result<u16> {
+    ) -> anyhow::Result<TokenAmount> {
         self.update_submitters(store, submitter)?;
+        self.total_submission_weight += submitter_weight.clone();
```
Comment (adlrocha) [inline gateway/src/cron.rs:147]: Not a blocker (and I may be wrong) but I think that if you borrow here you save having to clone `submitter_weight`.
```suggestion
        self.total_submission_weight += &submitter_weight;
```

File: gateway/src/cron.rs [hunk 3]
```diff
@@ -189,13 +201,13 @@ impl CronSubmission {
...
+        //     MOST_VOTED                 THRESHOLD     TOTAL_SUBMISSIONS  TOTAL_WEIGHT
         //
-        // we see MOST_VOTED is smaller than THRESHOLD, TOTAL_SUBMISSIONS and TOTAL_VALIDATORS, if
-        // the potential extra votes any vote can obtain, i.e. TOTAL_VALIDATORS - TOTAL_SUBMISSIONS,
+        // we see MOST_VOTED is smaller than THRESHOLD, TOTAL_SUBMISSIONS and TOTAL_WEIGHT, if
+        // the potential extra votes any vote can obtain, i.e. TOTAL_WEIGHT - TOTAL_SUBMISSIONS,
         // is smaller than or equal to the potential extra vote the most voted can obtain, i.e.
         // THRESHOLD - MOST_VOTED, then consensus will never be reached, no point voting, just abort.
-        if threshold - most_voted_count >= total_validators - self.total_submissions {
+        if threshold - most_voted_weight >= total_weight - self.total_submission_weight.clone() {
```
Comment (adlrocha) [inline gateway/src/cron.rs:210]: Like above, I think this works if I am not wrong.
```suggestion
        if threshold - most_voted_weight >= total_weight - &self.total_submission_weight {
```

**Review by adlrocha — APPROVED** [review]
LGTM. Let's have a new PR with unit tests and I will do a final pass over everything before merging. Thanks.

File: gateway/src/cron.rs (hunk 2 above)
Comment (cryptoAtwill) [reply gateway/src/cron.rs:147]: yep, it works. I will update.

**Pull request merged**